### PR TITLE
feat(sdk): added the validation for verifyTxn for trx

### PR DESCRIPTION
**Pull request by sachinBitgo**
COIN-3579

📋 Summary
This PR implements comprehensive transaction verification functionality for TRON (TRX) transfers in the BitGo SDK. The changes add robust validation for TransferContract transactions to ensure transaction integrity and security.

🔧 Changes Made

Core Implementation

- Added verifyTransaction method in trx.ts to validate transaction parameters against the prebuild transaction
- Implemented validateTransferContract for specific validation of native TRX transfers
- Created generateRawDataHex utility in utils.ts for proper protobuf encoding of transaction data
- Added getHexFromBase64 helper for base64 to hex conversion

Key Features

✅ Parameter validation: Ensures required fields (txParams, txPrebuild, wallet) are present
✅ Address validation: Validates owner and destination addresses using Base58 encoding
✅ Amount verification: Confirms transaction amounts match between client request and prebuild
✅ Address matching: Verifies destination addresses match recipient addresses
✅ Contract structure validation: Ensures proper TransferContract parameter structure

Error Handling

- Comprehensive error messages for missing or invalid parameters
- Specific validation for different failure scenarios
- Debug logging for transaction verification failures

🧪 Testing

- Added comprehensive test suite (verifyTransaction.ts)
- Parameter validation tests: Missing txParams, wallet, txPrebuild scenarios
- Contract validation tests: Valid and invalid TransferContract scenarios
- Edge case testing: Missing addresses, amounts, and mismatched values
- Updated test resources with corrected address values

### CODE REVIEW

File: modules/sdk-coin-trx/src/trx.ts [hunk 1]
```diff
@@ -250,6 +251,75 @@ export class Trx extends BaseCoin {
   }
 
   async verifyTransaction(params: VerifyTransactionOptions): Promise<boolean> {
+    const { txParams, txPrebuild, wallet } = params;
+    assert(txParams, 'missing txParams');
+    assert(txPrebuild && wallet, 'missing txPrebuild');
+    assert(txPrebuild.txHex, 'missing txHex in txPrebuild');
```
Comment (mohammadalfaiyazbitgo) [inline modules/sdk-coin-trx/src/trx.ts]: nit: prefer explicit `Error` over assert messages.

File: modules/sdk-coin-trx/src/trx.ts [hunk 2]
```diff
@@ -250,6 +251,75 @@ export class Trx extends BaseCoin {
...
+        throw new Error('Number of contracts is greater than 1.');
+      }
+
+      const contract = txJson.raw_data.contract[0];
+
+      if (contract.type === 'TransferContract') {
+        return this.validateTransferContract(contract, txParams);
+      } else {
+        return true;
+      }
```
Comment (mohammadalfaiyazbitgo) [inline modules/sdk-coin-trx/src/trx.ts]: there's no other transaction type to verify, like a token enablement, or a consolidation?

**Review by mohammadalfaiyazbitgo — CHANGES_REQUESTED** [review]

File: modules/sdk-coin-trx/src/trx.ts (hunk 2 above)
Comment (sachinBitgo) [reply modules/sdk-coin-trx/src/trx.ts]: Yes, but might me in future other contract type can come like vote, freeze etc.

File: modules/sdk-coin-trx/src/trx.ts (hunk 1 above)
Comment (sachinBitgo) [reply modules/sdk-coin-trx/src/trx.ts]: Resolved

**Review by mohammadalfaiyazbitgo — DISMISSED** [review]

File: modules/sdk-coin-trx/src/trx.ts [hunk 3]
```diff
@@ -250,6 +250,84 @@ export class Trx extends BaseCoin {
...
+    }
+
+    const rawTx = txPrebuild.txHex;
+    const txBuilder = getBuilder(this.getChain()).from(rawTx);
+    const tx = await txBuilder.build();
+    const txJson = tx.toJson();
+
+    try {
+      if (!txJson.raw_data || !txJson.raw_data.contract || txJson.raw_data.contract.length !== 1) {
+        throw new Error('Number of contracts is greater than 1.');
```
Comment (kamleshmugdiya) [inline modules/sdk-coin-trx/src/trx.ts]: why are you throwing error instead of returning `false`?

Comment (sachinBitgo) [reply modules/sdk-coin-trx/src/trx.ts]: this method is not expecting any return value

Comment (kamleshmugdiya) [reply modules/sdk-coin-trx/src/trx.ts]: the return type is boolean of `verifyTransaction` func.

Comment (kamleshmugdiya) [reply modules/sdk-coin-trx/src/trx.ts]: if it's not expecting any return value, then the return type can be `void`

Comment (sachinBitgo) [reply modules/sdk-coin-trx/src/trx.ts]: correct @kamleshmugdiya but, it is not expecting at the caller. That's why need to throw error.
image

File: modules/sdk-coin-trx/src/trx.ts [hunk 4]
```diff
@@ -250,6 +250,84 @@ export class Trx extends BaseCoin {
...
+
+      if (contract.type === 'TransferContract') {
+        return this.validateTransferContract(contract, txParams);
+      } else {
+        return true;
+      }
+    } catch (e) {
+      console.debug('TSS transaction verification failed:', e.message);
+      throw e;
+    }
```
Comment (kamleshmugdiya) [inline modules/sdk-coin-trx/src/trx.ts]: no need of try catch.


```suggestion
      if (!txJson.raw_data || !txJson.raw_data.contract || txJson.raw_data.contract.length !== 1) {
        console.debug('TSS transaction verification failed:', e.message);
        throw new Error('Number of contracts is greater than 1.');
      }

      const contract = txJson.raw_data.contract[0];

      if (contract.type === 'TransferContract') {
        return this.validateTransferContract(contract, txParams);
      }
      return true;
    }
```

File: modules/sdk-coin-trx/src/trx.ts (hunk 3 above)
Comment (kamleshmugdiya) [reply modules/sdk-coin-trx/src/trx.ts]: understood.
in that case, there is no need to add try catch.
if you are only only throwing the error again from the catch block.

**Review by kamleshmugdiya — DISMISSED** [review]
lgtm

**Review by nvjsr — DISMISSED** [review]

**Review by mohammadalfaiyazbitgo — APPROVED** [review]

**Review by nvjsr — APPROVED** [review]
lgtm

**Review by kamleshmugdiya — APPROVED** [review]
